src/interpolation_methods.jl

```diff
@@ -1,13 +1,18 @@
 # Linear Interpolation
+function __interpolate(u1, u2, t1, t2, t)
+    θ = (t - t1)/(t2 - t1)
+    (1 - θ)*u1 + θ*u2
+end
 function _interpolate(A::LinearInterpolation{<:AbstractVector}, t::Number)
     idx = max(1, min(searchsortedlast(A.t, t), length(A.t) - 1))
     t1, t2 = A.t[idx], A.t[idx+1]
     u1, u2 = A.u[idx], A.u[idx+1]
-    t == t1 && return u1 # Return exact value if no interpolation needed (eg when NaN at t2)
-    t == t2 && return u2 # ... (eg when NaN at t1)
-    θ = (t - t1)/(t2 - t1)
-    return (1 - θ)*u1 + θ*u2
-  end
+    # For type stability of fast paths:
+    T = Core.Compiler.return_type(__interpolate, Tuple{typeof(u1), typeof(u2), typeof(t1), typeof(t2), typeof(t)})
+    t == t1 && return T(u1) # Return exact value if no interpolation needed (eg when NaN at t2)
+    t == t2 && return T(u2) # ... (eg when NaN at t1)
+    __interpolate(u1, u2, t1, t2, t)
```

Comment:
It's best to avoid `Core.Compiler.return_type`. It can cause extreme slowdowns when cases where the inputs aren't inferred. Usually, the way this is handled is to compute the general return value and use it for the special cases, i.e. to compute

```
θ = (t - t1)/(t2 - t1)
r = (1 - θ)*u1 + θ*u2
```

before the branches and then have

```julia
t == t1 && return oftype(r, u1)
```

Reply:
Yeah, sounds good.  The fast path ends up being no faster anyways and it rarely occurs with interpolations.